internal/dataplane/parser/errors.go

```diff
@@ -0,0 +1,75 @@
+package parser
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TranslationError represents an error occurring during translating Kubernetes objects into Kong ones.
+// It can be associated with one or more Kubernetes objects.
+type TranslationError struct {
+	causingObjects []client.Object
+	reason         string
+}
+
+// NewTranslationError creates a TranslationError with a reason that should be a human-readable explanation
+// of the error reason, and an causingObjects slice that specifies what objects have caused the error.
+func NewTranslationError(reason string, causingObjects ...client.Object) (TranslationError, error) {
+	if reason == "" {
+		reason = "unknown"
+	}
+	if len(causingObjects) < 1 {
+		return TranslationError{}, fmt.Errorf("no causing objects specified, reason: %s", reason)
+	}
```

Comment:
I wonder about that, we failed to create an error, thus we return an error. Can we change that to

```
return TranslationError{
		causingObjects: []client.Object{},
		reason:         reason,
	}
```

instead and get rid of second returned value?

Reply:
The rationale behind it is I consider `TranslationError` a value object that should itself be responsible for the validity of the data it is accepting. Thus, its constructor can return a regular error which is a signal that the data that was passed to it was invalid. 

Leaving `causingObjects: []client.Object{}` could be harmful as we expect at least one object will be causing the translation error. Validating it here gives us a guarantee that any `TranslationError` constructed with its constructor function will satisfy this requirement. Further down, it allows assuming we are working with a valid instance of `TranslationError` and not validating the data that belongs to it in other places.

Reply:
After renaming it to Failure, my comment no longer holds true. It's not less confusing than returning an auxiliary error on error construction